### PR TITLE
[REEF-1355] Adding extensibility to the YarnReefDotNetClient for Mult…

### PR DESCRIPTION
**Pull request by shulmanb**
…i Runtime

This addresses the issue by:
1) Allowing specifying launcher class
2) Allowing pluging in different App/Job parameteres serializers
3) Allowing specifying additional configuration for a job driver

JIRA:
  [REEF-1355](https://issues.apache.org/jira/browse/REEF-1355)

### CODE REVIEW

**Comment by shulmanb** [comment]
@afchung @markusweimer This is the second part of REEF-1054 that makes the YarnDotNETReefClient extensible. Any way for me to send PRs that depend on this one?

**Comment by markusweimer** [comment]
> Any way for me to send PRs that depend on this one?

Not that I am aware off.

File: lang/cs/Org.Apache.REEF.Client/YARN/Parameters/AdditionalDriverAppConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Client.YARN.Parameters
+{
+    [NamedParameter("Additional configuration that is needed for the config serialziation.", defaultValue: " ")]
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/Parameters/AdditionalDriverAppConfiguration.cs]: typo

File: lang/cs/Org.Apache.REEF.Client/YARN/WindowsYarnJobCommandProvider.cs [hunk 2]
```diff
@@ -66,6 +68,7 @@ internal sealed class WindowsYarnJobCommandProvider : IYarnJobCommandProvider
             _driverStdoutFilePath = driverStdoutFilePath;
             _driverStderrFilePath = driverStderrFilePath;
             _driverMaxPermSizeMB = driverMaxPermSizeMB;
+            _launcherClassName = string.IsNullOrWhiteSpace(launcherClassName) ? LauncherClassName : launcherClassName;
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/WindowsYarnJobCommandProvider.cs]: Why not just use a default?

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 3]
```diff
@@ -93,10 +99,16 @@ public void Submit(JobRequest jobRequest)
                 _driverFolderPreparationHelper.PrepareDriverFolder(jobRequest.AppParameters, localDriverFolderPath);
 
                 // prepare configuration
-                var paramInjector = TangFactory.GetTang().NewInjector(jobRequest.DriverConfigurations.ToArray());
+                if (!string.IsNullOrWhiteSpace(_additionalConfig))
+                {
+                    AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: Inject it

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 4]
```diff
@@ -171,6 +183,7 @@ public IJobSubmissionResult SubmitAndGetJobStatus(JobRequest jobRequest)
                     .Set(YarnCommandProviderConfiguration.DriverStderrFilePath, jobParameters.StderrFilePath.Value);
             }
 
+            // injector shoudld contain the IYarnJobCommandProvider implementation if other than default should be used
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: should

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 5]
```diff
@@ -60,15 +64,17 @@ public sealed class YarnREEFDotNetClient : IREEFClient
             IJobResourceUploader jobResourceUploader,
             REEFFileNames fileNames,
             IJobSubmissionDirectoryProvider jobSubmissionDirectoryProvider,
-            YarnREEFDotNetParamSerializer paramSerializer)
+            YarnREEFDotNetParamsSerializer paramsSerializer,
+            [Parameter(typeof(AdditionalDriverAppConfiguration))] string additionalConfig)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: Why not use `IConfigurationProvider`?

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFDotNetAppParamsSerializer.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Interface;
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    /// <summary>
+    /// Application Parameters serializer for yarn
+    /// </summary>
+    [DefaultImplementation(typeof(YarnREEFDotNetAppParamSerializer))]
+    internal interface IYarnREEFDotNetAppParamsSerializer
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnREEFDotNetAppParamsSerializer.cs]: Why do you need a separate interface?

**Comment by afchung** [comment]
@shulmanb done with a first pass.

**Comment by shulmanb** [comment]
closing this PR as it needs more thinking.

**Pull request closed without merging**